src/ui/mod.rs

```diff
@@ -1,27 +1,31 @@
 //! The ui module should contain all user interaction code. This allows your other code to focus on
 //! game logic rather than handling inputs/outputs.
-use anyhow::{ Result, bail, Context };
-use std::io;
-
-/// Prompts the user for a new move
-pub fn get_move<R, W>(mut reader: R, mut writer: W) -> Result<(usize, usize)>
-where
-    R: io::BufRead,
-    W: io::Write,
-{
-    let mut resp = String::new();
-    writer
-        .write_all(b"Enter your move:\n")
-        .with_context(|| "Failed to write")?;
-    reader.read_line(&mut resp).with_context(|| "Failed to readline")?;
-    // #[cfg(test)]
-    // This should only run when `cargo test`, but for some reason doesn't work
-    // writer.write_all(resp.as_bytes()).expect("Failed to write");
-
-    parse_move_position(resp.trim().to_ascii_uppercase())
+
+mod raw_mode_ui;
+mod text_ui;
+pub use raw_mode_ui::RawModeUi;
+pub use text_ui::TextUi;
+
+use crate::board::Board;
+use anyhow::{bail, Result};
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum UserAction {
+    Move(usize, usize),
+    Quit,
+    Noop,
+}
+
+pub trait UserInterface {
+    /// User input, which will be passed to the controller
+    fn input(&mut self) -> Result<UserAction>;
+
+    /// View the model
+    fn view(&mut self, board: &Board) -> Result<()>;
 }
 
-fn parse_move_position(mv: String) -> Result<(usize, usize)> {
+fn parse_move_position(mv: &str) -> Result<(usize, usize)> {
```

Comment:
Tests need to be updated since changing to string slice

Reply:
Fixed, added. Added more tests caught a bug